pkg/updatechecker/updatechecker.go

```diff
@@ -513,13 +516,49 @@ func autoDeploy(opts CheckForUpdatesOpts, clusterID string, autoDeploy apptypes.
 		return nil
 	}
 
+	if err := waitForPreflightsToFinish(opts.AppID, versionToDeploy.Sequence); err != nil {
+		return errors.Wrap(err, "not able to auto-deploy due to failed preflight check")
+	}
+
 	if err := deployVersion(opts, clusterID, appVersions, versionToDeploy); err != nil {
 		return errors.Wrapf(err, "failed to deploy sequence %d with version label %s", versionToDeploy.Sequence, versionToDeploy.VersionLabel)
 	}
 
 	return nil
 }
 
+func waitForPreflightsToFinish(appID string, sequence int64) error {
+	err := wait.PollImmediate(2*time.Second, 15*time.Minute, func() (bool, error) {
+		versionStatus, err := store.GetStore().GetDownstreamVersionStatus(appID, sequence)
+		if err != nil {
+			return false, errors.Wrap(err, "failed get status")
+		}
+		if versionStatus != storetypes.VersionPendingPreflight {
+			return true, nil
+		}
+		return false, nil
+	})
+	if err != nil {
+		return errors.Wrap(err, "failed to poll for preflights results")
+	}
+	// refetch latest results
+	preflightResult, err := store.GetStore().GetPreflightResults(appID, sequence)
+	if err != nil {
+		return errors.Wrap(err, "failed to fetch preflight results")
+	}
+
+	var parsedResults preflight.UploadPreflightResults
+	if err = json.Unmarshal([]byte(preflightResult.Result), &parsedResults); err != nil {
+		return errors.Wrap(err, "failed to parse preflight results")
+	}
+
+	if parsedResults.Errors != nil {
+		return errors.Wrap(errors.New(preflightResult.Result), "preflight errors")
+	}
```

Comment:
preflight result doesn't account for failed analyzer

preflightresult.results[].isFail=true, the deployments should be blocked.

+ when there is no preflight spec in the app, we may need to test the edge case